### ISSUE TITLE
Error when retrieving an irregular multidimensional array

### ISSUE THREAD

**Issue by d-tomerini** [aiidateam/aiida-core#3411]
AiiDA does not allow to retrieve irregularly shaped arrays (list of lists with different lengths) when saved to disk. Numpy saves them as pickles and does not allow retrieval without an `` allow_pickle=True`` flag.

This results in being able to save them to disk, but not to load them afterwards.

Test code by @sphuber 

```
import numpy
from aiida.orm import ArrayData
data = [[1, 1, 1], [2, 2], [3, 3, 3, 3]]
array = ArrayData()
ndarray = numpy.array([numpy.array(element) for element in data])

array.set_array('scf_convergence', ndarray)
array.store()
array.get_array('scf_convergence') 
```

**Comment by greschd** [comment]
This seems to apply for all ``numpy`` arrays with ``dtype=object``. For example, ``np.array([1, 'a', dict()])`` has the same behavior.

But, I'd be quite sceptical of allowing ``pickle`` as a long-term storage format. Unpickling breaks as soon as one of the classes used in the data structure changes names, for example. It should probably be detected already when calling ``set_array``, though.

In this case, wouldn't it be possible to just use a ``List`` instead?

**Comment by sphuber** [comment]
The thing I don't understand is why it would have a `dtype=object` in the example I gave. I explicitly make all lists numpy arrays. So every element should be a numpy type. Even the integers are `numpy.int64`. So why is it than still storing as a pickle? I even tried to call `numpy.save` manually with `allow_pickle=False` and that also fails with the same exception.

**Comment by greschd** [comment]
Yes, numpy arrays are supposed to be homogeneous -- that's necessary for them to be contiguous in memory with O(1) random access.

The example given doesn't produce a "single" numpy array, but rather an "array of arrays" that contains the lower arrays as objects.

**Comment by greschd** [comment]
Fun side note: Strings in numpy arrays are automatically truncated (to the size set when the array is created).

```python
In [32]: x = np.array(['a', 'a', 'a'])                                                                                  
In [33]: x[0] = 'test'                                                                                                  
In [34]: x                                                                                                              
Out[34]:                                                                                                                
array(['t', 'a', 'a'],
      dtype='<U1')
```

**Comment by sphuber** [comment]
OK that makes sense. The reason we stumbled upon this is because we would like to parse the SCF convergence accuracy to the parsed nodes. Since for calculations with multiple SCF cycles each cycle can have a different number of steps, this list of lists will not be a homogeneous matrix. We were planning of adding it to the `output_trajectory` output node which has type `TrajectoryData`, which stores its content as numpy files on disk. This was giving us the exception. So as @d-tomerini already identified, there are a few options:
* pad the nested lists to make the matrix homogeneous
* use two one-dimensional arrays, one with actual values and one to mark which belongs to what snapshot
* actually use pickles

Main drawback of the first is that it is not space efficient and not sure what kind of padding values would be possible. Zeroes or negative values should work, but both are a bit weird of course.
Second solution adds more complexity to use the data.

Using a `List` is possible, but it adds another output node (not necessarily a problem, but to me it makes more sense in the `TrajectoryData`) but a bigger problem is that the `List` stores its contents in the database. Since the data can be quite big and does not really need to be queryable, I am not sure this is an ideal solution.

**Comment by greschd** [comment]
Mandatory watching regarding pickles: https://www.youtube.com/watch?v=7KnfGDajDQw

The pickle problems probably aren't as bad for numpy since it's pretty stable, but it still feels wrong to me.

How big are those arrays typically? If they're really big, you could also add each cycle as a separate array with indexed names.. a bit ugly due to the int -> str -> int conversion, but it should be pretty stable.

**Comment by d-tomerini** [comment]
Hi @greschd 

Not big. At their worst, each one of the 1-d array can be up to ~100, for each of the relax steps (less than a 100?). But it adds up for every single calculation

**Comment by greschd** [comment]
Hmm, then maybe the second option (index array) makes sense? I would guess it's the most space efficient.

**Comment by greschd** [comment]
While ``pickle`` might not be too troublesome in the case of nested ``numpy`` arrays, I think it's dangerous (in that people might create data that will be unreadable later on) to allow pickle in ``ArrayData``. So, I think it would be wise to put ``allow_pickle=False`` in the **saving** method. If we absolutely need the option, maybe exposing it to ``set_array`` would be good.

With data sharing, are we at all concerned about untrusted data? Because unpickling untrusted data allows arbitrary code execution - which is why numpy switched to ``allow_pickle=False`` in the loading method.

**Comment by sphuber** [comment]
This morning I was thinking exactly about your last comment. Originally I thought, what is the problem with using pickles. People are generating and reading the data themselves. But then off course with data sharing this opens a nasty attack vector. Someone could create an export archive containing an `ArrayData` whose repository contains a pickled array with objects that will trigger code execution. They only need to get someone to import it and call the `get_array` method on it. Even though it is not very likely, this could potentially be very problematic so I think we should not even expose the functionality.

**Comment by greschd** [comment]
I think this can be closed since #3434 is merged, and the remaining work discussed here is happening in ``aiida-quantumespresso``.
